src/main/java/com/puppycrawl/tools/checkstyle/utils/XpathUtil.java

```diff
@@ -0,0 +1,67 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2018 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.utils;
+
+import java.util.Arrays;
+import java.util.List;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * Contains utility methods for xpath.
+ *
+ */
+public final class XpathUtil {
+
+    /** List of token types which support value attribute. */
+    private static final List<Integer> TOKEN_TYPES_WITH_VALUE_ATTRIBUTE = Arrays.asList(
+            TokenTypes.IDENT, TokenTypes.STRING_LITERAL, TokenTypes.CHAR_LITERAL,
+            TokenTypes.NUM_LONG, TokenTypes.NUM_INT, TokenTypes.NUM_DOUBLE, TokenTypes.NUM_FLOAT);
```

Comment:
I will try my best to explain as clear as possible)
Reasons why I did this limitation:

1) I think `@value` should be used for values, not AST token values, but java variable values. I could not find any extra case where I would need this attribute except for querying tokens by value.

2) The `@value` attribute takes its value from `DetailAst.getText()` method. 
So it is possible that `@value` attribute will be the same as token name:
`//CLASS_DEF[@value='CLASS_DEF']//MODIFIERS[@value='MODIFIERS']`
Or almost the same:
```
//LITERAL_VOID[@value='void']
//RCURLY[@value='}']
```
I think `@value` attribute is useless for those queries, it does not carry any information, it does not help to get more precise results, it does nothing.

3) This limitation is used also by `XpathQueryGenerator`.  With current logic it would generate a very long queries with `@value` attribute in every node in the path.
For example:
```
/CLASS_DEF[@text='InputXpathQueryGenerator' and @value='CLASS_DEF']
/OBJBLOCK[@value='OBJBLOCK']/METHOD_DEF[@text='callSomeMethod' and @value='METHOD_DEF']
/MODIFIERS[@value='MODIFIERS']/LITERAL_PUBLIC[@value='public']
```

What do you think?

Reply:
done